go-controller/pkg/node/default_node_network_controller.go

```diff
@@ -1440,6 +1448,116 @@ func (nc *DefaultNodeNetworkController) WatchNamespaces() error {
 	return err
 }
 
+func (nc *DefaultNodeNetworkController) WatchNodes() error {
+	_, err := nc.retryNodes.WatchResource()
+	return err
+}
+
+// addOrUpdateNode handles creating routes or nftables rules for each node to handle PMTUD
+func (nc *DefaultNodeNetworkController) addOrUpdateNode(node *corev1.Node) error {
+	hostCIDRs := nc.Gateway.GetHostCIDRs()
+	for _, address := range node.Status.Addresses {
+		if address.Type != corev1.NodeInternalIP {
+			continue
+		}
+		nodeIP := net.ParseIP(address.Address)
+		if nodeIP == nil {
+			continue
+		}
+		var existingHostSubnet *net.IPNet
+		for _, hostCIDR := range hostCIDRs {
+			if utilnet.IsIPv6CIDR(hostCIDR) != utilnet.IsIPv6(nodeIP) {
+				continue
+			}
+			if hostCIDR.Contains(nodeIP) {
+				existingHostSubnet = hostCIDR
+				break
+			}
+		}
+
+		// subnet exists on our node
+		if existingHostSubnet != nil {
+			// create MTU lock routes
+			mask := 32
+			nlFamily := netlink.FAMILY_V4
+			if utilnet.IsIPv6(nodeIP) {
+				mask = 128
+				nlFamily = netlink.FAMILY_V6
+			}
+			routes, err := util.GetNetLinkOps().RouteList(nil, nlFamily)
+			if err != nil {
+				return fmt.Errorf("unable to list host routes: %w", err)
+			}
+			linkIndex := -1
+			existingRoute := netlink.Route{}
+			for _, route := range routes {
+				if existingHostSubnet.Contains(route.Dst.IP) {
+					linkIndex = route.LinkIndex
+					existingRoute = route
+					break
+				}
+			}
+			if linkIndex == -1 {
+				return fmt.Errorf("unable to find suitable host route for node %s with IP: %s, existing host subnet: %s",
+					node.Name, nodeIP, existingHostSubnet)
+			}
+			_, fullCIDR, err := net.ParseCIDR(fmt.Sprintf("%s/%d", nodeIP, mask))
+			if err != nil {
+				return fmt.Errorf("unable to parse CIDR for node %s: %v", node.Name, err)
+			}
+			existingRoute.Dst = fullCIDR
+			link, err := util.GetNetLinkOps().LinkByIndex(linkIndex)
+			if err != nil {
+				return err
+			}
+			mtu := link.Attrs().MTU
+			existingRoute.MTU = mtu
+			existingRoute.MTULock = true
+			klog.Infof("Adding route for remote node: %s, IP: %s, locked MTU: %d", node.Name, nodeIP, mtu)
+			if err := nc.routeManager.Add(existingRoute); err != nil {
+				return err
+			}
+		}
+		// TODO(trozet): else case...create nftables rules to block egress of ICMP needs frag
+
+	}
+	return nil
+}
+
+func (nc *DefaultNodeNetworkController) deleteNode(node *corev1.Node) {
```

Comment:
wheres the sync func upon restart? how you handling missed Node delete events?

Reply:
ah its in a later commit with route manager.

Reply:
actually now that i pulled out route manager stuff, i realize i need a sync for nftables with pmtud...i'll add it. Thanks for catching it